applications/matter_bridge/src/bridged_devices_creator.cpp

```diff
@@ -0,0 +1,384 @@
+		delete provider;
+	}
+
+	return err;
+}
+#else
+CHIP_ERROR BridgedDeviceCreator::CreateDevice(int deviceType, const char *nodeLabel, chip::Optional<uint8_t> index,
+					      chip::Optional<uint16_t> endpointId)
+{
+#if defined(CONFIG_BRIDGED_DEVICE_SIMULATED)
+	BridgedDeviceDataProvider *provider = CreateSimulatedProvider(deviceType);
+	/* The device is simulated, so it can be added immediately. */
+	return AddMatterDevice(deviceType, nodeLabel, provider, index, endpointId);
+#else
+	return CHIP_ERROR_NOT_IMPLEMENTED;
+#endif /* CONFIG_BRIDGED_DEVICE_SIMULATED */
+}
+#endif /* CONFIG_BRIDGED_DEVICE_BT */
+
+CHIP_ERROR BridgedDeviceCreator::RemoveDevice(int endpointId)
+{
+	uint8_t index;
+	uint8_t count = 0;
+	uint8_t indexes[BridgeManager::kMaxBridgedDevices] = { 0 };
+	size_t indexesCount = 0;
+	CHIP_ERROR err = BridgeManager::Instance().RemoveBridgedDevice(endpointId, index);
+
+	if (CHIP_NO_ERROR != err) {
+		LOG_ERR("Failed to remove bridged device");
+		return err;
+	}
+
+	if (!BridgeStorageManager::Instance().LoadBridgedDevicesIndexes(indexes, BridgeManager::kMaxBridgedDevices,
+									indexesCount)) {
+		LOG_ERR("Failed to load stored bridged device indexes");
+		return CHIP_ERROR_INTERNAL;
+	}
+
+	/* Find the required index on the list, remove it and move all following indexes one position earlier. */
+	bool indexFound = false;
+	for (auto i = 0; i < static_cast<int>(indexesCount); i++) {
+		if (indexes[i] == index) {
+			indexFound = true;
+		}
+
+		if (indexFound && ((i + 1) < BridgeManager::kMaxBridgedDevices)) {
+			indexes[i] = indexes[i + 1];
+		}
+	}
+
+	if (indexFound) {
+		indexesCount--;
+	} else {
+		return CHIP_ERROR_NOT_FOUND;
+	}
+
+	/* Update the current indexes list. */
+	if (!BridgeStorageManager::Instance().StoreBridgedDevicesIndexes(indexes, indexesCount)) {
+		LOG_ERR("Failed to store bridged devices indexes.");
+		return CHIP_ERROR_INTERNAL;
+	}
+
+	if (!BridgeStorageManager::Instance().LoadBridgedDevicesCount(count)) {
+		LOG_ERR("Failed to load bridged devices count.");
+		return CHIP_ERROR_INTERNAL;
+	}
+
+	if (!BridgeStorageManager::Instance().StoreBridgedDevicesCount(count - 1)) {
+		LOG_ERR("Failed to store bridged devices count.");
+		return CHIP_ERROR_INTERNAL;
+	}
+
+	if (!BridgeStorageManager::Instance().RemoveBridgedDeviceEndpointId(index)) {
+		LOG_ERR("Failed to remove bridged device endpoint id.");
+		return CHIP_ERROR_INTERNAL;
+	}
+
+	/* Ignore error, as node label may not be present in the storage. */
+	BridgeStorageManager::Instance().RemoveBridgedDeviceNodeLabel(index);
+
+	if (!BridgeStorageManager::Instance().RemoveBridgedDeviceType(index)) {
+		LOG_ERR("Failed to remove bridged device type.");
+		return CHIP_ERROR_INTERNAL;
+	}
+
+#ifdef CONFIG_BRIDGED_DEVICE_BT
+	if (!BridgeStorageManager::Instance().RemoveBtAddress(index)) {
+		LOG_ERR("Failed to remove bridged device Bluetooth address.");
+		return CHIP_ERROR_INTERNAL;
+	}
```

Comment:
Isn't there any chance to refactor it so that we don't need to have storage read followed by storage write for each component? I guess that after device is removed with `RemoveBridgedDevice`, the `BrigdeManager` is aware of the current configuration because it's stored internally in the DeviceMap, so maybe we only need to somehow get this information directly and overwrite the current storage?

Reply:
Yeah, reads before writes are actually needed only to know the current list of indexes and their count. I think it should be possible to refactor the code to make `BridgeManager` keep and expose such a list.

Reply:
I refactored the code to avoid loading before storing. Please re-review.